--- shmenu
+++ shmenu-b
@@ -5,7 +5,7 @@
 exec 3>&1
 [[ -c /dev/tty ]] && exec </dev/tty >/dev/tty
 
-trap 'printf "\e[H%${COLUMNS}s\e8\e[?7h"' EXIT
+trap 'printf "\e[${LINES}H%${COLUMNS}s\e8\e[?7h"' EXIT
 trap 'key=redraw' WINCH
 
 printf '\e7\e[?7l'
@@ -30,7 +30,8 @@
 		index=(0 0)
 	}
 
-	printf '\e[?25l\e[H\e[30;41m %s%*s\e[m' \
+	printf '\e[?25l\e[%sH\e[30;41m %s%*s\e[m' \
+		"$LINES" \
 		"$text" \
 		$(( longest - ${#text} + 4 )) "${index[2]:+<} "
 
@@ -44,8 +45,8 @@
 				(( length += ${#matches[i]} + 2 ))
 				(( max_items++ ))
 			else
-				printf '\e[30;41m\e[;%sH%*s%s\e[m' \
-					"$length" \
+				printf '\e[30;41m\e[%sH%*s%s\e[m' \
+					"$LINES;$length" \
 					$(( COLUMNS - length - 1 )) "" \
 					"> "
 				(( length = COLUMNS ))
@@ -62,15 +63,15 @@
 
 	(( ${#matches[@]} )) && {
 		matches[sel]=$'\e[D\e[0;7m '${matches[sel]}
-		printf '\e[;%sH' $(( longest + 6 ))
+		printf "\e[$LINES;$(( longest + 6 ))H"
 		printf '\e[30;41m %s \e[m' \
 			"${matches[@]:index:max_items-index}"
 		matches[sel]=${matches[sel]#$'\e[D\e[0;7m '}
 	}
 
-	printf '\e[30;41m%*s\e[m\e[?25h\e[;%sH' \
+	printf '\e[30;41m%*s\e[m\e[?25h\e[%sH' \
 		$(( COLUMNS - length )) "" \
-		$(( ${#text} + 2 ))
+		"$LINES;$(( ${#text} + 2 ))"
 
 	until read -t 0.05 -rsN 1; do
 		[[ $key == redraw ]] && break
@@ -87,6 +88,7 @@
 
 	case $key in
 		$'\n' | $'\r')
+			printf "\e[${LINES}H%${COLUMNS}s"
 			printf '\e8'
 			if (( ${#matches[@]} )); then
 				printf '%s\n' "${matches[sel]}"
