--- shmenu
+++ vol
@@ -1,6 +1,8 @@
 #!/usr/bin/env bash
+#
+# Sets the volume to the selected percentage using amixer.
 
-mapfile -t items
+items=({1..100})
 shopt -s checkwinsize
 exec 3>&1
 exec </dev/tty >/dev/tty || exit
@@ -54,6 +56,12 @@ while (:); do
 			fi
 		done
 
+		[[ ! -v text_cache ]] && {
+			[[ $(amixer -M get Master) =~ .*\[([0-9]+)%\].* ]]
+			(( sel = BASH_REMATCH[1] - 1 ))
+			text_cache=
+		}
+
 		(( sel && sel >= max_items )) && {
 			index+=("$max_items")
 			index[0]=$max_items
@@ -88,16 +96,9 @@ while (:); do
 
 	case $key in
 		$'\n' | $'\r')
-			printf '\e8'
-
-			if (( ${#matches[@]} )); then
-				: "${matches[sel]}"
-			else
-				: "$text"
-			fi
-			eval "printf '%s\n' $_" >&3
-
-			printf '\e7'
+			(( ${#matches[@]} )) && {
+				amixer -M set Master "${matches[sel]}%"
+			} &>/dev/null
 			exit
 			;;
 		$'\e')
